### PR TITLE
[ALLUXIO-1799] Add unit test for alluxio.collections.DirectedAcyclicGraph

### PR DESCRIPTION
**Pull request by peisun1986**
https://alluxio.atlassian.net/browse/ALLUXIO-1799

Add unit test for alluxio.collections.DirectedAcyclicGraph

### CODE REVIEW

**Comment by gpang** [comment]
Hi @peisun1986 , I think this PR is also including commits from one of your other PRs.

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Tests the {@link DirectedAcyclicGraph} class.
+ */
+public class DirectedAcyclicGraphTest {
```
Comment (apc999) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: could you mark this class as `final`? In this code base, we prefer to explicitly annotate those classes final when they are not expected to have subclass.

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    /*
+    * Initialize.
+     */
+    @Before
+    public void before() {
+        mGraph = new DirectedAcyclicGraph<>();
+    }
+
+    /*
+    * Test some simple graphs.
```
Comment (apc999) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: NIT: by alluxio javadoc convention, `Test`-> `Tests`

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        for (int i = 1; i <= 6; ++i) {
+            toSort.add(i);
+        }
+
+        List<Integer> result;
+        Set<Integer> seen = new HashSet<>();
+
+        result = mGraph.sortTopologically(toSort);
+        for (Integer i : result) {
+            if (i == 1) {
```
Comment (apc999) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: Alluxio codebase use 2-space indent.

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java (hunk 2 above)
Comment (peisun1986) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: Done

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java (hunk 3 above)
Comment (peisun1986) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: Done

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java (hunk 1 above)
Comment (peisun1986) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: Done.

**Comment by peisun1986** [comment]
Comments addressed. Thanks.

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+   */
+  @Test
+  public void simpleGraphTest() {
+    // Empty graph.
+    Assert.assertTrue(mGraph.getRoots().isEmpty());
+    Assert.assertTrue(mGraph.sortTopologically(new HashSet<Integer>()).isEmpty());
+
+    // One node.
+    mGraph.add((Integer) (1), new ArrayList<Integer>());
+    Assert.assertEquals(mGraph.getRoots().size(), 1);
```
Comment (gpang) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: I think for `assertEquals()` the expected value should be first, and the actual value you are testing comes second. If so, please update all the relevant asserts to that parameter ordering.

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    List<Integer> result;
+    Set<Integer> seen = new HashSet<>();
+
+    result = mGraph.sortTopologically(toSort);
+    for (Integer i : result) {
+      if (i == 1) {
+        Assert.assertTrue(seen.contains((Integer) (5)));
+      } else if (i == 2) {
+        Assert.assertTrue(seen.contains((Integer) (1)));
```
Comment (gpang) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: This means if we see node 2, we must have seen its parents, right? If so, why must we have seen 1 as well?

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    // Sort part of the graph.
+    toSort.clear();
+    toSort.add((Integer) (1));
+    toSort.add((Integer) (2));
+    seen.clear();
+
+    result = mGraph.sortTopologically(toSort);
+    for (Integer i : result) {
+      if (i == 2) {
+        Assert.assertTrue(seen.contains((Integer) (1)));
```
Comment (gpang) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: Why does 2 depend on 1? In the construction, 2 only depends on 3 and 4. 3 and 4 don't depend on anything.

**Comment by gpang** [comment]
Thanks for the test, @peisun1986 ! I left a few comments and questions.

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java (hunk 5 above)
Comment (peisun1986) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: good catch. I added 1->2 in the comment (line 61) but didn't add it here. fixed.

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java (hunk 6 above)
Comment (peisun1986) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: same as above

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java (hunk 4 above)
Comment (peisun1986) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: Done

**Comment by peisun1986** [comment]
Comments addressed. Thank you, Gene

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    parents.add((Integer) (1));
+    parents.add((Integer) (3));
+    parents.add((Integer) (4));
+    mGraph.add((Integer) (2), parents);
+    parents.clear();
+    mGraph.add((Integer) (6), parents);
+
+    // Sort the whole graph.
+    Set<Integer> toSort = new HashSet<>();
+    for (int i = 1; i <= 6; ++i) {
```
Comment (gpang) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: NIT: I think technically, the style is to use i++, for primitives. (different from the google c++ style...)

**Comment by gpang** [comment]
I left a very minor comment. Otherwise looks good!

File: core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java (hunk 7 above)
Comment (peisun1986) [inline core/common/src/test/java/alluxio/collections/DirectedAcyclicGraphTest.java]: Done.

**Comment by peisun1986** [comment]
Fixed.

**Comment by gpang** [comment]
LGTM

**Comment by gpang** [comment]
Jenkins, test this please

**Pull request merged**